Simplify comparison between address of member expressions

hannes-steffenhagen-diffblue opened the pull request:
For a struct like `struct S { int x; };` and a `struct S s;`, this lets comparisons like `&s.x == &s.x` and `(int*)&s == &s.x` be simplified, also similar expressions work for unions now.

For the validity of this simplification see [C11 Draft Standard N1570](http://www.iso-9899.info/n1570.html), section 6.7.2.1, paragraphs 15 and 16.

hannes-steffenhagen-diffblue committed:
Simplify comparisons involving address of struct/union members

thk123 commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
   return true;
 }
 
+static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
```
Please add Doxygen comments to each function please

thk123 commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
 
+static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
+{
+  exprt result = op;
```
Think I commented on the original, but a `PRECONDITION(op.typ()==ID_pointer)` I suspect might make the parameter to this function a bit more obvious what it is meant to be. Also consider giving it a more descriptive name (e.g. `pointer_expression`)

thk123 commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+  return result;
+}
+
+// returns true if op is a symbol or an expression like x.y[constant].z
```
Add an extra `/` to turn into a doxygen comment!

thk123 commented on unit/util/simplify_expr.cpp:
```diff
@@ -145,3 +145,159 @@ TEST_CASE("Simplify Java short -> int -> short casts")
+
+  signedbv_typet int_type(32);
+
+  struct_typet simple_struct;
```
For readability, consider adding a comment showing in C like syntax what you are building:
```C
struct { int x; int y; } s;
union {int x; int y; } u;
```

hannes-steffenhagen-diffblue replied:
Oh right, I meant to do that but it got lost in refactoring. I will rectify.

thk123 commented on unit/util/simplify_expr.cpp:
```diff
@@ -145,3 +145,159 @@ TEST_CASE("Simplify Java short -> int -> short casts")
+  {
+    THEN("Equality should simplify to true")
+    {
+      auto equals = equal_exprt(
```
Again, add comment showing  the expression in C like syntax:
```C
&s.x == &s.x
```

thk123 commented on unit/util/simplify_expr.cpp:
```diff
@@ -145,3 +145,159 @@ TEST_CASE("Simplify Java short -> int -> short casts")
+          address_of_exprt(simple_struct_var), pointer_type(int_type)),
+        address_of_exprt(member_exprt(simple_struct_var, "y", int_type)));
+      REQUIRE(simplify_expr(notequals, ns) == true_exprt());
+    }
```
Consider adding extra tests for structs:

```C
&s.y == &s.y
```

thk123 reviewed (approved):
Looks great - some suggestions for comments and tests, but no need for re-review if they're all done

smowton commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+}
+
+// returns true if op is a symbol or an expression like x.y[constant].z
+static bool simplify_inequality_is_simple_structure_expression(const exprt &op)
```
I don't think this needs the prefix, especially since it's static. The `simplify_inequality_` just suggest it simplifies an inequality, which it doesn't.

smowton commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
   return true;
 }
 
+static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
```
How about just `look_through_typecasts`, then do the postcondition in the caller?

hannes-steffenhagen-diffblue replied:
Didn't know that was a thing. Will use that instead.

smowton replied:
That's not a reference to an existing function, rather a suggested name

hannes-steffenhagen-diffblue replied:
It does exist in `java_pointer_casts.cpp`, but as a static function. Didn't realise you just meant a rename.

smowton replied:
ha, coincidence. Factoring to a general `util/` helper is probably sensible.

smowton commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+    exprt lhs_offset = simplify_expr(lhs_descriptor.offset(), ns);
+    exprt rhs_offset = simplify_expr(rhs_descriptor.offset(), ns);
+    bool equal = lhs_descriptor.root_object() == rhs_descriptor.root_object() &&
+                 lhs_offset.is_constant() && lhs_offset == rhs_offset;
```
This looks dodgy since sometimes it really indicates we don't know if they're equal, but I think by `is_simple_structure_expression` above we always have `lhs_offset.is_constant()`? If so make that an invariant.

tautschnig replied:
I agree this is dodgy: If the offsets aren't constant, then we can only give a definitive answer when the `expr.id() == ID_equal && equal`.

hannes-steffenhagen-diffblue replied:
Does indeed look wrong, I moved the `lhs_offset.is_constant()` into an `INVARIANT` (we shouldn't even get to this point if they aren't constant, as @smowton pointed out).

smowton commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+  return expr.id() == ID_symbol;
+}
+
 bool simplify_exprt::simplify_inequality_address_of(exprt &expr)
```
Clearly this was misnamed to begin with, but suggest changing to `simplify_equality_or_inequality...` since the last line makes it clear we're handling both `ID_equal` and `ID_notequal` here.

smowton reviewed (commented):
Some specific comments, and I recommend extending the unit test to cover the non-constant offset case too (i.e. things that should *not* simplify).

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -420,37 +421,64 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+    simplify_inequality_is_simple_structure_expression(lhs) &&
+    simplify_inequality_is_simple_structure_expression(rhs))
+  {
+    object_descriptor_exprt lhs_descriptor, rhs_descriptor;
```
With this work, I'm not sure I see a reason for the first commit (or indeed the `simplify_inequality_address_of_prepare_op`)?

hannes-steffenhagen-diffblue replied:
I'm stripping the typecasts for `is_simple_structure_expression`. It's probably not necessary to have the refactoring as a separate commit, it was only that way because I did that first before making any changes, I can squash all of that into one commit if you prefer.

tautschnig replied:
Why is the call to `simplify_inequality_is_simple_structure_expression` even necessary in the first place?

Re squashing: possibly, but there will be room for further changes as the entire preparation before constructing the `object_descriptor_exprt` now seems unnecessary (whether in the original or in the refactored form).

hannes-steffenhagen-diffblue replied:
Regarding `is_simple_structure_expression`; This is there to check that the expression is something that is safe to simplify (as opposed to constructs like `&random_struct()->x == &random_struct()->x` - in that case, the `root_object`s would be `*random_struct()`, and the offsets would both be constant and equal, so without the `is_simple_structure_expression` with this logic the simplifier would (incorrectly) simplify this to `true`.

tautschnig reviewed (changes requested):
I'm happy about `object_descriptor_exprt`, but there are issues left as highlighted by comments by others and myself.

hannes-steffenhagen-diffblue commented:
@smowton Regarding the unit tests for nonconstant: This would only happen in case of address of array element. This could code deal with that, but address of array element doesn't actually get to this part of the code because it gets rewritten to `pointer + offset` before that; So any unit test I'd be writing for that case wouldn't actually test the code that I want to test here.

hannes-steffenhagen-diffblue commented:
@tautschnig I believe I've addressed all the comments here now, I understand you are busy right now, but would you mind taking another look when you next have the time?

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -11,6 +11,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <cassert>
 
 #include "c_types.h"
+#include "simplify_expr.h"
```
Why is this needed?

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,40 +418,78 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
-  assert(expr.type().id()==ID_bool);
-  assert(expr.operands().size()==2);
-  assert(expr.id()==ID_equal || expr.id()==ID_notequal);
+  exprt result = expr;
```
No need to construct a temporary `exprt`. Just keep a pointer. And then return a `const exprt &`.

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,40 +418,78 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+    INVARIANT(
+      lhs_offset.is_constant() && rhs_offset.is_constant(),
+      "Simple structure expressions should have constant offsets");
+    if(lhs_descriptor.root_object() != rhs_descriptor.root_object())
```
Do this check early, before simplifying offsets.

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,40 +418,78 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+    lhs_descriptor.build(lhs, ns);
+    rhs_descriptor.build(rhs, ns);
+    exprt lhs_offset = simplify_expr(lhs_descriptor.offset(), ns);
+    exprt rhs_offset = simplify_expr(rhs_descriptor.offset(), ns);
```
Use `simplify_rec` instead of using the public API.

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,40 +418,78 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
+/// returns true if op is a symbol or an expression like x.y[constant].z
+static bool is_simple_structure_expression(const exprt &op)
+{
+  exprt expr = op;
```
Use pointers. Don't create temporary `exprt` here.

tautschnig commented on src/util/simplify_expr_pointer.cpp:
```diff
@@ -417,40 +418,78 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
-       tmp1.op0().get(ID_identifier);
+  // strip the address-of operator
+  lhs = lhs.op0();
+  rhs = rhs.op0();
```
Now I'm wondering whether the following would also be equivalent to the below:

````c++
mp_integer lhs_offset = compute_pointer_offset(lhs, ns);
if(lhs_offset == -1)
  return true;

mp_integer rhs_offset = compute_pointer_offset(rhs, ns);
if(rhs_offset == -1)
  return true;

bool is_equal = lhs_offset == rhs_offset;
bool should_be_equal = expr.id() == ID_equal;
expr.make_bool(is_equal == should_be_equal);

return false;
````
